third_party/e2e/engine.cc

```diff
@@ -0,0 +1,91 @@
+#include "engine.h"
+
+#include <cstdint>
+
+#include "lib/translator/translation.h"
+#include "nvme_driver.h"
+#include "util.h"
+
+namespace {
+
+constexpr uint8_t kTimeout = 60;
+
+}  // namespace
+
+void SetEngineCallbacks(void) {
+  translator::SetDebugCallback(Print);
+  translator::SetAllocPageCallbacks(AllocPages, DeallocPages);
+}
+
+ScsiToNvmeResponse ScsiToNvme(unsigned char* cmd_buf, unsigned short cmd_len,
+                              unsigned long long lun, unsigned char* sense_buf,
+                              unsigned short sense_len, unsigned char* data_buf,
+                              unsigned short data_len, bool is_data_in) {
+  // Create translation object
+  translator::Translation translation;
+
+  // Package parameters and run translation begin
+  translator::Span<uint8_t> scsi_cmd(cmd_buf, cmd_len);
+  translator::Span<uint8_t> buffer(data_buf, data_len);
+  translator::BeginResponse begin_resp =
+      translation.Begin(scsi_cmd, buffer, lun);
+
+  if (begin_resp.status == translator::ApiStatus::kFailure) {
+    Print("Incorrect usage of Translation Library API");
+    ScsiToNvmeResponse resp = {
+        .return_code = static_cast<uint8_t>(scsi::Status::kTaskAborted),
+        .alloc_len = 0};
+    return resp;
+  }
+
+  if (begin_resp.alloc_len > data_len) {
+    Print(
+        "Specified allocation length exceeds buffer size. Possible malicious "
+        "request?");
+    ScsiToNvmeResponse resp = {
+        .return_code = static_cast<uint8_t>(scsi::Status::kTaskAborted),
+        .alloc_len = 0};
+    return resp;
+  }
+
+  // Grab NVMe cmds and call NVMe interface
+  translator::Span<const translator::NvmeCmdWrapper> nvme_wrappers =
+      translation.GetNvmeWrappers();
+  nvme::GenericQueueEntryCpl cpl_buf[nvme_wrappers.size()] = {};
```

Comment:
cpl?

Reply:
Looks like a shorthand for 'completion' with widespread use in the project.